Multiple Resource Interpreter (Chat with all FHIR records)

AdritRao opened the pull request:
<!--

This source file is part of the Stanford Biodesign for Digital Health open-source project

SPDX-FileCopyrightText: 2022 Stanford University and the project authors (see CONTRIBUTORS.md)

SPDX-License-Identifier: MIT

-->

# Multiple Resource Interpreter (Chat with all FHIR records)

## :recycle: Current situation & Problem
Currently, there is a single resource interpreter with individual chat interfaces for each FHIR resource. The goal is to enable a general conversation with all health records in a single interface.
Link to issue: https://github.com/StanfordBDHG/LLMonFHIR/issues/22

## :bulb: Proposed solution
A "multiple resource interpreter" is added to enable a general conversation with all health records. This interpreter takes an array of all FHIR resources and uses a separate prompt to allow the user to converse with all health records.

## :gear: Release Notes 
A button was added to the toolbar titled "Chat with all records". This takes the user to a chat interface that allows them to converse with all health records. The main additions to the code were:
1. [FHIRMultipleResourceInterpreter.swift](https://github.com/StanfordBDHG/LLMonFHIR/blob/multiple_resource_interpreter/LLMonFHIR/FHIR%20Display/FHIRMultipleResourceInterpreter.swift) (this is a modified interpreter which takes an array of FHIR resources and uses a separate `interpret_multiple_resources` prompt)
2. [MultipleResourceChat.swift](https://github.com/StanfordBDHG/LLMonFHIR/blob/multiple_resource_interpreter/LLMonFHIR/FHIR%20Display/MultipleResourceChat.swift) (this is the chat interface for multiple resources which the user can access through the "Chat with all records" button on the toolbar)
3. [FHIRResourcesView.swift](https://github.com/StanfordBDHG/LLMonFHIR/blob/multiple_resource_interpreter/LLMonFHIR/FHIR%20Display/FHIRResourcesView.swift) (modified so that it interprets all resources when the app loads and then enables the "Chat with all records" feature)

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on LLMonFHIR.xcodeproj/project.pbxproj:
```diff
@@ -100,6 +102,8 @@
 		2FE5DC4229EDD7F2004B9AB4 /* Binding+Negate.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = "Binding+Negate.swift"; sourceTree = "<group>"; };
 		2FE5DC4329EDD7F2004B9AB4 /* Bundle+Image.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = "Bundle+Image.swift"; sourceTree = "<group>"; };
 		2FE5DC4429EDD7F2004B9AB4 /* CodableArray+RawRepresentable.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; path = "CodableArray+RawRepresentable.swift"; sourceTree = "<group>"; };
+		4333523F2A5C95E00043A440 /* MultipleResourceChat.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; name = MultipleResourceChat.swift; path = "../../../../Downloads/LLMonFHIR-multiple_resource_interpreter 2/LLMonFHIR/FHIR Display/MultipleResourceChat.swift"; sourceTree = "<group>"; };
+		433352432A5C96090043A440 /* FHIRMultipleResourceInterpreter.swift */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.swift; name = FHIRMultipleResourceInterpreter.swift; path = "../../../../Downloads/LLMonFHIR-multiple_resource_interpreter 2/LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift"; sourceTree = "<group>"; };
```
Seems like the files were added with non-relative to the surrounding folder setting. You can change this in the Xcode "File Inspector" panel on the trailing side of the Xcode window. I think that "Relative to Group" should be the correct setting that we should aim for.

PSchmiedmayer reviewed (commented):
@AdritRao Thank you for the generation of the PR; this is be a great and important addition to the application! 🚀

It seems like the SwiftLint checks are currently failing due to some violations of the style guidelines, you can see the different annotations in the [PR diff](https://github.com/StanfordBDHG/LLMonFHIR/pull/23/files). I would recommend installing SwiftLint locally on your Mac so you can see the warnings right in your project: https://github.com/realm/SwiftLint#using-homebrew

In addition to that, it seems like the [build is failing ](https://github.com/StanfordBDHG/LLMonFHIR/actions/runs/5515149044/jobs/10055175978#step:17:281) due to some missing files. I have annotated the reason for that in the Xcode project diff that should resolve that issue as it seems like they are committed in the repo!

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift:
```diff
@@ -0,0 +1,92 @@
+        for try await chatStreamResult in chatStreamResults {
+            for choice in chatStreamResult.choices {
+                let previousInterpretation = interpretations ?? ""
+                interpretations = previousInterpretation + (choice.delta.content ?? "")
```
You can directly use a one-liner here:
```suggestion
                interpretations = (interpretations ?? "") + (choice.delta.content ?? "")
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift:
```diff
@@ -0,0 +1,92 @@
+    func chat(resources: [FHIRResource]) -> [Chat] {
+        var chat = [systemPrompt(forResources: resources)]
+        
+        if let interpretation: String? = interpretations {
+            if let interpretation = interpretation {
+                chat.append(Chat(role: .assistant, content: interpretation))
+            }
+        }
```
A nice suggestion to write this in a more compact fashion:
```suggestion
        if let interpretations = interpretations {
            chat.append(Chat(role: .assistant, content: interpretations))
        }
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/MultipleResourceChat.swift:
```diff
@@ -0,0 +1,70 @@
+import SwiftUI
+
+
+struct MultipleResourceChat: View {
```
The `MultipleResourceChat` looks like a very close copy to the existing view that displays the chat in a single resource. I have documented a general need for a generalized View here: https://github.com/StanfordSpezi/SpeziML/issues/17 . We don't need to do this in the same PR to bring this view to SpeziML, but would be good to ensure that we have a minimal code duplication here. Maybe let's refactor this in a `ResourceChatView` or something similar that contains all the information except the `NavigationStack` and the `navigationTitle` modifier. This would allow us to reuse the view for both use cases and then only need to package that view in a NavigationStack and add a `navigationTitle` modifier.

AdritRao replied:
I added a ``` ResourceChatView ``` which handles the chat interface for both single and multiple resources.

PSchmiedmayer commented on LLMonFHIR/LLMonFHIRDelegate.swift:
```diff
@@ -31,7 +32,7 @@ class LLMonFHIRDelegate: SpeziAppDelegate {
                 [
                     HKClinicalType(.allergyRecord),
-                    HKClinicalType(.clinicalNoteRecord),
+//                    HKClinicalType(.clinicalNoteRecord),
```
If you have updated the deployment target to iOS 16.4, you should be able to re-add the `.clinicalNoteRecord` data type.

PSchmiedmayer commented on LLMonFHIR/Resources/Localizable.strings:
```diff
@@ -118,6 +118,29 @@ Immediately return an interpretation to the user, starting the conversation.
+
+Tell the user that they can ask any question about their health records and then create a short list of the main categories of health records of the user which you have access to.
+
+Immediately return an interpretation to the user, starting the conversation.
+The initial interpretation should be a short and simple summary with the following specifications:
+1. Overall summary of all health records
+2. Middle school reading level
+3. End with a question asking user if they have any questions. Make sure that this question is not generic but specific to their health records.
+Do not introduce yourself at the beginning, and start with your interpretation.
+Make sure your response is in the same language the user writes to you in.
+The tense should be present.
```
Nice use of some clear instructions here! 🚀

AdritRao replied:
Thank you!

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -46,14 +70,49 @@ struct FHIRResourcesView: View {
+                            }
+                        }) {
+                            if interpretingMultipleResources {
+                                ProgressView("Interpreting...")
```
Please use localized strings using the `Localizable.strings` file when you add new strings to the application.

AdritRao replied:
I removed this portion of the code because I added a "Chat with all records" button to the top of the list which uses a ```OnboardingActionsView``` section.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -46,14 +70,49 @@ struct FHIRResourcesView: View {
+                            if interpretingMultipleResources {
+                                ProgressView("Interpreting...")
+                            } else {
+                                Text("Chat with All Records")
```
Please use localized strings using the `Localizable.strings` file when you add new strings to the application.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/MultipleResourceChat.swift:
```diff
@@ -0,0 +1,70 @@
+        NavigationStack {
+            ChatView($chat, disableInput: $gettingAnswer)
+                
+                .navigationTitle("All FHIR Resources")
```
Please use localized strings using the `Localizable.strings` file when you add new strings to the application.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/MultipleResourceChat.swift:
```diff
@@ -0,0 +1,70 @@
+    var body: some View {
+        NavigationStack {
+            ChatView($chat, disableInput: $gettingAnswer)
+                
```
Please avoid empty lines when chaining view modifier.
```suggestion
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -46,14 +70,49 @@ struct FHIRResourcesView: View {
                 }
                 .toolbar {
+                    ToolbarItem(placement: .navigationBarTrailing) {
+                        Button(action: {
+                            if !interpretingMultipleResources {
+                                showMultipleResourcesChat.toggle()
+                            } else {
+                            }
+                        }) {
+                            if interpretingMultipleResources {
+                                ProgressView("Interpreting...")
+                            } else {
+                                Text("Chat with All Records")
+                            }
+                        }
+                    }
```
While I like to first focus on the functionality of the app, I would already have one first suggestion on how to improve the UI a bit to feel a bit more like the current UI design of the app and how Apple designs their apps. While the toolbar placement works for one item, multiple trailing toolbar items are not often used in mobile applications. 

I would therefore suggest placing the button a bit more prominent and centered in the UI, maybe something like this?
<img width="370" alt="Screenshot 2023-07-11 at 8 42 29 PM" src="https://github.com/StanfordBDHG/LLMonFHIR/assets/28656495/4a406e70-8d35-4e54-868e-1d57e2f8b66a">

This could, e.g. be achieved using the `OnboardingActionsView` from the SpeziOnboarding package and placed above the list of resource types (currently line [53](https://github.com/StanfordBDHG/LLMonFHIR/pull/23/files#diff-8c1ada103658e4da8f3b8d89b13392bd55097d5ef5e0e5aa6dea39dbd4a8b733R53)) 

```swift
Section {
    OnboardingActionsView(
        "CHAT_WITH_ALL_RESOURCES",
        action: {
            await interpretMultipleResources()
            showMultipleResourcesChat.toggle()
        }
    )
        .padding(-20)
}
```

I would also suggest a slightly different flow of events in the app.
Currently the `interpretMultipleResources()` function is called right after the view is loaded. I would suggest that we only call this function when the user wants to chat about all resources and e.g. presses the button like shown in the button implementation above. 
The best solution would be if we already show the chat view right after the user pressed the button, then load the query in the background, and then populate the chat. That might not be needed in this version of the PR and might play well with the following feature suggestion: https://github.com/StanfordSpezi/SpeziML/issues/15

AdritRao replied:
Sounds great! I removed the button from the tool bar and added the ```OnboardingActionsView``` button to the top of the list. I used the same action which you provided so that ```interpretMultipleResources()``` is only called when the button is tapped.

PSchmiedmayer reviewed (changes requested):
@AdritRao Thank you so much for the PR and the improvements! A great step in the right direction. Thank you for fixing the build and the SwiftLint issues. 

I have added some comments throughout the PR to share some suggestions to improve the code in the next iteration. Let me know what you think about the suggestions and if you have any follow-up questions! 🚀

AdritRao commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,17 +7,41 @@
     
+    private enum FHIRMultipleResourceInterpreterConstants {
+        static let storageKey = "FHIRMultipleResourceInterpreter.Cache"
+    }
```
I updated the code so that it uses this storage key to access the saved interpretation from UserDefaults and only runs the initial interpretation if no previous interpretation was performed.

AdritRao commented:
Hi Paul, thank you for your comments on the PR! I went through your comments and updated the code in the latest commit. I made the following changes:
1. Updated the [.swiftlint.yml](https://github.com/StanfordBDHG/LLMonFHIR/pull/23/files/18428c8f46a853e9d9aa8288a183a90d143ee86e#diff-7109ad4ca9c6e705a8da290e7cb2898a679b0f8f1f4712d082dee256230d6a1f) to include all default tests except attributes (and fixed the code so that it is now passing all of the tests)
2. Created a single ``` ResourceChatView.swift ``` which handles the chat UI for both single and multiple resources
3. Updated the UI so that instead of a toolbar button, there is a ```OnboardingActionsView``` section with a "Chat with all records" button at the top of the list (and so that ```interpretMultipleResources()``` is only run when the button is tapped). 
4. I also made many minor changes based on your comments such as re-adding the .clinicalNoteRecord, making the code more compact, and removing the typealias.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift:
```diff
@@ -0,0 +1,87 @@
+class FHIRMultipleResourceInterpreter<ComponentStandard: Standard>: DefaultInitializable, Component, ObservableObject, ObservableObjectProvider {
+    @Dependency private var localStorage: LocalStorage
+    @Dependency private var openAIComponent = OpenAIComponent()
+    
```
Very minor change: We normally add two empty lines between different types of elements in a type, e.g. two empty lines between stored properties and computed properties in that case. The rest of the file correctly uses this pattern 🚀
```suggestion
    
    
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift:
```diff
@@ -0,0 +1,87 @@
+        var chat = [systemPrompt(forResources: resources)]
+        
+        if let interpretation = interpretation {
+                    chat.append(Chat(role: .assistant, content: interpretation))
```
Please only indent new lines in new scopes by 4 characters:
```suggestion
           chat.append(Chat(role: .assistant, content: interpretation))
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,31 +7,39 @@
 import SwiftUI
 
+private enum FHIRMultipleResourceInterpreterConstants {
+        static let storageKey = "FHIRMultipleResourceInterpreter.Cache"
+    }
```
Please only indent new lines in new scopes by 4 characters:
```suggestion
    static let storageKey = "FHIRMultipleResourceInterpreter.Cache"
}
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,31 +7,39 @@
 import SpeziOpenAI
 import SwiftUI
 
+private enum FHIRMultipleResourceInterpreterConstants {
```
This type is no longer needed if we move the cache checking functionality to the `FHIRMultipleResourceInterpreter`.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,31 +7,39 @@
 
 struct FHIRResourcesView: View {
-    @EnvironmentObject var fhirStandard: FHIR
+    typealias MultipleResourceInterpretation = String
```
This type is no longer needed if we move the cache checking functionality to the `FHIRMultipleResourceInterpreter`.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,31 +7,39 @@
     @State var showSettings = false
+    @State var showMultipleResourcesChat = false
+    @State var interpretingMultipleResources = false
+    @State var error: String?
```
The error is currently not used anywhere, we could also just print the error when it is emitted.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,31 +7,39 @@
-    
     
+    @EnvironmentObject var fhirMultipleResourceInterpreter: FHIRMultipleResourceInterpreter<FHIR>
+    @EnvironmentObject var localStorage: LocalStorage<FHIR>
```
This type is no longer needed if we move the cache checking functionality to the `FHIRMultipleResourceInterpreter`.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -51,9 +59,39 @@ struct FHIRResourcesView: View {
+                .sheet(isPresented: $showMultipleResourcesChat) {
+                    resourceChatView
+                }
+                .task {
+                    allResourcesArray = await fhirStandard.resources
+                    await interpretMultipleResources()
+                }
```
You can move this functionality to the closure in the `OnboardingActionsView`. We can also set the `allResourcesArray` in there as well.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -51,9 +59,39 @@ struct FHIRResourcesView: View {
+        ResourceChatView(
+            chat: fhirMultipleResourceInterpreter.chat(resources: allResourcesArray),
+            title: "All FHIR Resources"
+         )
```
Please indent elements by 4 spaces.
```suggestion
        )
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -118,6 +165,22 @@ struct FHIRResourcesView: View {
     }
+    
+    private func interpretMultipleResources() async {
+        guard let cache: MultipleResourceInterpretation = try? localStorage.read(storageKey: FHIRMultipleResourceInterpreterConstants.storageKey)
+        else {
+            interpretingMultipleResources = true
+            do {
+                try await fhirMultipleResourceInterpreter.interpretMultipleResources(resources: fhirStandard.resources)
+            } catch {
+                self.error = error.localizedDescription
+            }
+
+            interpretingMultipleResources = false
+
+            return
+        }
```
I would suggest to move the check if the cache is found to the `FHIRMultipleResourceInterpreter`. As you already load the cached results in the `interpretation` property you also don't need to use the cache there, something like 
```swift
guard interpretation == nil else {
     return
}
```
in that method should be sufficient.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/ResourceChatView.swift:
```diff
@@ -0,0 +1,71 @@
+struct ResourceChatView: View {
+    @Environment(\.dismiss) var dismiss
+    @EnvironmentObject var openAPIComponent: OpenAIComponent<FHIR>
+    @EnvironmentObject var fhirResourceSummary: FHIRResourceSummary<FHIR>
+    @EnvironmentObject var fhirStandard: FHIR
```
You can remove those two properties, they are not used here.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/ResourceChatView.swift:
```diff
@@ -0,0 +1,71 @@
+    @State var gettingAnswer = false
+    
+    let title: String
+
```
Very minor change: We normally add two empty lines between different types of elements in a type, e.g. two empty lines between stored properties and computed properties in that case.
```suggestion


```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/ResourceChatView.swift:
```diff
@@ -0,0 +1,71 @@
+                }
+        }
+    }
+    
```
Very minor change: We normally add two empty lines between different types of elements in a type, e.g. two empty lines between stored properties and computed properties in that case.
```suggestion


```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/MultipleResourceChat.swift:
You can remove this file and all the types in there as we now use the shared `ResourceChatView`.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/InspectResourceChat.swift:
You can remove this file and all the types in there as we now use the shared `ResourceChatView`.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/ResourceChatView.swift:
I would suggest to rename this to `OpenAIChatView.swift`. It doesn't really have any knowledge about FHIR resources so this might be a more applicable name.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/ResourceChatView.swift:
```diff
@@ -0,0 +1,71 @@
+import SwiftUI
+
+
+struct ResourceChatView: View {
```
I would suggest to rename this to `OpenAIChatView`. It doesn't really have any knowledge about FHIR resources so this might be a more applicable name.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift:
```diff
@@ -0,0 +1,87 @@
+
+    
+    private func systemPrompt(forResources resources: [FHIRResource]) -> Chat {
+        var allJSONResources = ""
+        
+        for resource in resources {
+            allJSONResources = allJSONResources + "\n" + resource.compactJSONDescription
+        }
```
A nice trick to save some computation power, memory allocation, and make this code use some functional programming in Swift is to use `map` and `joined` to simplify the code here: 
```suggestion
        let allJSONResources = resources.map(\.compactJSONDescription).joined(separator: "\n")
```

PSchmiedmayer commented on LLMonFHIR.xcodeproj/project.pbxproj:
```diff
@@ -599,10 +603,10 @@
 				ASSETCATALOG_COMPILER_GLOBAL_ACCENT_COLOR_NAME = AccentColor;
 				CODE_SIGN_ENTITLEMENTS = "LLMonFHIR/Supporting Files/LLMonFHIR.entitlements";
 				CODE_SIGN_IDENTITY = "Apple Development";
-				CODE_SIGN_STYLE = Automatic;
+				CODE_SIGN_STYLE = Manual;
 				CURRENT_PROJECT_VERSION = 1;
 				DEVELOPMENT_ASSET_PATHS = "";
-				DEVELOPMENT_TEAM = 637867499T;
+				DEVELOPMENT_TEAM = "";
```
Please ensure that you have enabled "Automatic Code Signing" + selected "None" as your development team in the Test Configuration for all targets:
<img width="1178" alt="Screenshot 2023-07-12 at 6 33 03 PM" src="https://github.com/StanfordBDHG/LLMonFHIR/assets/28656495/6ce3ed28-481a-4fc5-97b5-210057cfcf5c">

PSchmiedmayer commented on LLMonFHIR.xcodeproj/project.pbxproj:
```diff
@@ -664,7 +668,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				CURRENT_PROJECT_VERSION = 1;
-				DEVELOPMENT_TEAM = 637867499T;
+				DEVELOPMENT_TEAM = Y65AD5HVCG;
```
The release configuration should stay the same using @PSchmiedmayer's development team.
<img width="1178" alt="Screenshot 2023-07-12 at 6 36 12 PM" src="https://github.com/StanfordBDHG/LLMonFHIR/assets/28656495/aac87fd2-ab57-43e4-a4b2-2f061336e868">

PSchmiedmayer reviewed (changes requested):
Thank you for all the changes and improvements, we are very close to merging the PR! Added a few comments here and there that should be addressed before we merge it 🚀

Some of the diff changes that should not be there in the Xcode project are only really editable using a text editor (Xcode project files are really not perfect), I will push a small commit to set the signing settings back to the previous state.
You will need to do a `git pull` to get the latest changes and ensure that you can commit on top of them when incorporating the feedback.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRMultipleResourceInterpreter.swift:
```diff
@@ -0,0 +1,88 @@
+    }
+    
+    func interpretMultipleResources(resources: [FHIRResource]) async throws {
+        guard interpretation != nil else {
+            let chatStreamResults = try await openAIComponent.queryAPI(withChat: [systemPrompt(forResources: resources)])
+            
+            for try await chatStreamResult in chatStreamResults {
+                for choice in chatStreamResult.choices {
+                    let previousInterpretation = interpretation ?? ""
+                    interpretation = (interpretation ?? "") + (choice.delta.content ?? "")
+                }
+            }
+            return
+        }
```
A general code suggestion that is also applicable beyond this change: A general thing that we aim for in Swift code is keeping the "golden" path (the one where we are in a non-error state) outside of nested statements. This allows you to make the code a bit more readable and guard statements are a perfect way to enforce that we exit early in case if an error.

Therefore I would suggest to change the code like this:
```suggestion
        guard interpretation == nil else {
            return
        }
        
        let chatStreamResults = try await openAIComponent.queryAPI(withChat: [systemPrompt(forResources: resources)])
        
        for try await chatStreamResult in chatStreamResults {
            for choice in chatStreamResult.choices {
                let previousInterpretation = interpretation ?? ""
                interpretation = (interpretation ?? "") + (choice.delta.content ?? "")
            }
        }
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -7,31 +7,31 @@
 //
 
 import ModelsR4
+import SpeziLocalStorage
```
We don't use this import statement anymore:
```suggestion
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/FHIRResourcesView.swift:
```diff
@@ -118,6 +155,20 @@ struct FHIRResourcesView: View {
+        }
+
+        interpretingMultipleResources = false
+        
+        return
```
You can remove the return statement here:
```suggestion
```

PSchmiedmayer commented on LLMonFHIR.xcodeproj/project.pbxproj:
```diff
@@ -798,7 +802,7 @@
 				ASSETCATALOG_COMPILER_GLOBAL_ACCENT_COLOR_NAME = AccentColor;
 				CODE_SIGN_ENTITLEMENTS = "LLMonFHIR/Supporting Files/LLMonFHIR.entitlements";
 				CODE_SIGN_IDENTITY = "Apple Development";
-				CODE_SIGN_STYLE = Automatic;
+				CODE_SIGN_STYLE = Manual;
```
Please ensure that the release settings are the same as before, probably the easiest way to do this is to change this line in a text editor.

PSchmiedmayer commented on LLMonFHIR.xcodeproj/project.pbxproj:
```diff
@@ -844,12 +848,10 @@
 				ASSETCATALOG_COMPILER_GLOBAL_ACCENT_COLOR_NAME = AccentColor;
 				CODE_SIGN_ENTITLEMENTS = "LLMonFHIR/Supporting Files/LLMonFHIR.entitlements";
 				CODE_SIGN_IDENTITY = "Apple Development";
-				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Distribution";
 				CODE_SIGN_STYLE = Manual;
 				CURRENT_PROJECT_VERSION = 1;
 				DEVELOPMENT_ASSET_PATHS = "";
-				DEVELOPMENT_TEAM = "";
-				"DEVELOPMENT_TEAM[sdk=iphoneos*]" = 637867499T;
+				DEVELOPMENT_TEAM = 637867499T;
```
Please ensure that the release settings are the same as before, probably the easiest way to do this is to change this line in a text editor.

PSchmiedmayer reviewed (changes requested):
@AdritRao Thank you for all the changes!

I have added some smaller comments and unresolved some comments that are not yet addressed from previous reviews and might have just slipped through a previous push. If there is any reason why the don't make any sense feel free to note down the reason under the comment 🚀

PSchmiedmayer commented:
@AdritRao I just played around with the app and one thing that I found that we missed is adding the new prompt to the `SettingsView`. This should be a very minimal change as we just need do add a new settings screen in the prompt section in this section:
```swift
Section("SETTINGS_PROMPTS") {
    NavigationLink(value: SettingsDestinations.promptSummary) {
        Text("SETTINGS_PROMPTS_SUMMARY")
    }
    NavigationLink(value: SettingsDestinations.promptInterpretation) {
        Text("SETTINGS_PROMPTS_INTERPRETATION")
    }
}
```
Everything else is already so well abstracted that just adding a new `NavigationLink` should do all the work.

AdritRao commented:
Hi @PSchmiedmayer, thank you for your comments! I have updated the code based on your feedback and added a new ```NavigationLink``` to the ```SettingsView``` so that the user can modify the multiple resource interpretation prompt. Please let me know if you have any other suggestions. Thank you!

PSchmiedmayer reviewed (approved):
Thank you @AdritRao, a great job with the PR!

There is one last conversation that requires resolution and we can then merge the PR!
https://github.com/StanfordBDHG/LLMonFHIR/pull/23#discussion_r1260507757

Thank you for all the work and the great improvements, looking forward to the next contributions and all the stuff we discussed in the last meetings that bring LLMonFHIR to the next level 🚀
